src/Microsoft.Kiota.Cli.Commons/Http/Headers/BaseHeadersStore.cs

```diff
@@ -0,0 +1,92 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Kiota.Cli.Commons.Http.Headers;
+
+/// <summary>
+/// Base class for the <see cref="IHeadersStore"/> interface.
+/// This type exists so all child classes can share the headers parsing logic.
+/// </summary>
+public abstract class BaseHeadersStore : IHeadersStore
+{
+    /// <inheritdoc />
+    public abstract IEnumerable<KeyValuePair<string, ICollection<string>>> Headers();
+
+    /// <inheritdoc />
+    /// <remarks> 
+    /// Passing an empty collection of headers to this function has the same
+    /// effect as calling <see cref="Drain"/>
+    /// </remarks>
+    public virtual IEnumerable<KeyValuePair<string, ICollection<string>>> SetHeaders(ICollection<string> headers)

```

Comment:
defensive programming missing at numerous places

Reply:
Are there cases you can see that aren't mentioned by the code docs?

Reply:
in case a null value is passed

Reply:
The input is not nullable. i.e. `ICollection<string>?`

Reply:
Mhhh...I thought the point of NRT was to help avoid having to check for nulls. Let me see what I can do to work around that. I'm thinking about trying to use overload resolution maybe? 🤔

Reply:
it's a dev time feature, it doesn't enforce anything at runtime (through AOP or what not) Maybe the [DisallowNullAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.disallownullattribute) does something at runtime, but it's not clear. The easiest here is probably to add a ArgumentNullException.ThrowIfNull guard

Reply:
There's a null check here already. By this time, we know headers will not be null. https://github.com/microsoft/kiota-cli-commons/blob/5aae5b44e254519d5abbc085f8cca4ce80409ad6/src/Microsoft.Kiota.Cli.Commons/Extensions/CommandBuilderExtensions.cs#L127-L130

Would you prefer the null check is moved to this function?

Reply:
> Maybe the [DisallowNullAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.disallownullattribute) does something at runtime, but it's not clear

It seems like the attribute doesn't do anything in a null aware context. The weird thing is that when I add null checks, I get hints to remove them from Roslyn...

Reply:
Doesn't matter, the method I commented on is public and could be used in another context

Reply:
I changed the API to accept a null collection. I think that's a better experience than throwing. thoughts?

Reply:
I don't have a strong opinion either way. And this would be an opinion thing anyway. Tolerating out of bound values or failing as early as possible to provide the most precise feedback to the user of an API have been two school of thoughts for decades. I'd say I've usually experienced dynamic type languages to do the former and the strictly type languages to do the latter.

Reply:
I think dynamically typed languages need the runtime checks since there's no type system that will notify a developer when they pass in the wrong thing. I can provide a string, number, boolean, null to any function in javascript so runtime checks for type are necessary. However, some libraries don't even add them and are very successful. As an example, check out D3js's source. they don't perform some checks because they have documented what the APIs allow.

Back to static languages, code will fail to compile if you pass the wrong type provided there are compile time checks. This gives us 1 layer of protection. Another layer is the debug asserts to ensure invariants are honored during development time. Adding runtime checks aggressively everywhere where compile-time checks have been done has little benefit in my opinion. It greatly reduces the benefit of having a release build configuration in the first place.